### ISSUE TITLE
Migrate this repo to a LLVM repository

### ISSUE THREAD

**Issue by ChristianKuehnel** [google/llvm-premerge-checks#186]
Right now it's in a Google repo, but as this should be part of LLVM, we should move it there.

Discuss what to do with the issues: Keep them on GitHub or move to bugzilla?

Next step: Discuss with LLVM Ops team about this change.

**Comment by ldionne** [comment]
Strong +1 on this -- this is LLVM's CI infrastructure, it should be part of the monorepo. IMO this is fairly high priority, too.

**Comment by joker-eph** [comment]
I read the bug entry here as "move this repo under the llvm-project", not "inside the monorepo". The latter seems more controversial, as for example the prior haven't gone in this direction (see https://github.com/llvm/llvm-zorg for example)

**Comment by ChristianKuehnel** [comment]
I don't really have a strong opinion of where to put it as long as it's clearly a part of LLVM.

**Comment by ldionne** [comment]
> I read the bug entry here as "move this repo under the llvm-project", not "inside the monorepo". The latter seems more controversial, as for example the prior haven't gone in this direction (see https://github.com/llvm/llvm-zorg for example)

`llvm-zorg` was in its own repository before we made a move to the monorepo for other projects (is that correct?), so that could explain why they opted for the status quo and didn't move it to the monorepo. Perhaps it's just that nobody bothered doing it.

FWIW, I think it's pretty important to make this move. I have a preference for putting it in the monorepo since it would be easy to see changes go through (e.g. herald rules and stuff), but I care mostly about moving this under the LLVM umbrella. This is probably the most critical piece of infrastructure for libc++, libc++abi and libunwind, so I think it makes sense to stop considering this as a one-off project now.

**Comment by joker-eph** [comment]
>  (e.g. herald rules and stuf

zorg is using Phabricator I believe: you can still have your herald rules and everything working as usual.

Having it as a separate repo also allows to define your own policy independently of the monorepo (pull-requests, GitHub issues, etc.)

**Comment by metaflow** [comment]
Agree, I will create a proposal to move this under LLVM

**Comment by metaflow** [comment]
@ldionne @joker-eph do you know what steps to follow to add a new repo under LLVM?

**Comment by ChristianKuehnel** [comment]
AFAIK there is not formal process for this. We have this (with some links) on the backlog of the IWG: https://github.com/llvm/llvm-iwg/issues/51

**Comment by joker-eph** [comment]
Yeah you just need someone with admin permission to create it right now, I usually emailed Tom/Mike/Anton in the past.

**Comment by ChristianKuehnel** [comment]
Just found this in the developer policy: https://llvm.org/docs/DeveloperPolicy.html#adding-an-established-project-to-the-llvm-monorepo

**Comment by ldionne** [comment]
Sorry for going MIA after the comment above.

I think we should move it into the monorepo so that we can see infrastructure changes going in just like the rest of code changes. I wouldn't create a separate repo for this.

**Comment by joker-eph** [comment]
We already land our infrastructure in a separate repo:  https://github.com/llvm/llvm-zorg  ; I'm not sure why this infrastructure is fundamentally different from the other docker images and testing scripts?

**Comment by aaupov** [comment]
From the perspective of a project joining monorepo (BOLT), I agree with Mehdi that this infrastructure appears redundant to buildbot configurations in llvm-zorg. End result is that we need to set up two parts of testing infrastructure in two different repos. That's not a huge problem, but definitely is an overhead.
On the topic, I think it would make sense to at least move this repo to llvm organization on github, with eventual convergence with llvm-zorg.

**Comment by ldionne** [comment]
FWIW my opinion is that llvm-zorg should *also* be part of the monorepo. Anyway, we might not get to a consensus on that, but at least let's move it under the llvm organization on GitHub (I think we all agree on that bit).

**Comment by metaflow** [comment]
@ldionne et al.

I don't have any preference to move to either of a) llvm-project b) llvm-zorg c) completely new repo under github/llvm.

One thing I would like to support is a testing and review process of commits:

1) as changes might affect cluster load significantly and also visible in many reviews.
2) I believe it makes sense for project that encourages pre-submit checks and peer-reviews to follow such practices.

I am also considering separating cluster management parts from checks. We still need to give people chance to inspect  e.g. Docker containers to reproduce builds and fix setup. Docker seems to be a good common ground to define common OS-es. Meanwhile GCP and Kubernetes setup scripts are tied to how current setup is running and also should be more restrictive to writes.

One argument for not taking (a) is that that we might want to run test scripts from a version different than sources under test. It's still possible to just check out a separate version nearby and use only scripts but sounds like a overhead.

I am planning to contact llvm-zorg and check what commit and testing policies they have and whether they will be OK to have pre-merge checks.

**Comment by metaflow** [comment]
Hi Aaron, thank you for the interest. While I would like to see that moved
somewhere under LLVM org, there are few points addressed, e.g. who is going
to maintain the infa from the community side.
also I don't see a lot of people eager to contribute to the premerges (and
the only fact that stops them is the wrong repo location). org is also not
very eager to e.g. hire a contractor for that. If everything will go fine I
think that the best chance of this thing to evolve is to start moving
towards PRs and GitHub actions.
what do you think?

--Mikhail

**Comment by joker-eph** [comment]
> If everything will go fine I think that the best chance of this thing to evolve is to start moving towards PRs and GitHub actions.

I share this sentiment: if this migration is not trivial (a few hours...), it is probably better at this point to focus the little human power available on the move out of Phabricator to GitHub PR!

**Comment by AaronBallman** [comment]
I don't think wanting the precommit CI repo to live within the project it tests is a reason to switch to github PRs (especially given that there is not clear consensus in the community to switch to GitHub PRs). The observation that not many people want to work on premerge is precisely why I am asking to shift it into the llvm repo -- almost nobody knows this repo exists over here, so when things fail, people give up pretty quickly after not finding how they can help. I've fielded at least two questions in the past few weeks on "where does precommit CI live?" and the follow-up question is always "why is it there???". Of course, whether they'd actually make changes if it was in the llvm project somewhere, who knows.

I'm rather ignorant of the inner workings of precommit CI though -- I take it there are issues making it a Bad Idea to have the repo hosted within llvm-project but not make any changes to the infrastructure it runs on? (e.g., are there concerns the LLVM community would suddenly add too much load onto Google's servers? If so... could we protect the repo so that a Google code owner has to sign off on commits to that folder?)

**Comment by joker-eph** [comment]
> I don't think wanting the precommit CI repo to live within the project it tests is a reason to switch to github PRs (especially given that there is not clear consensus in the community to switch to GitHub PRs).

I thought the move to PRs was already decided and we’re just months away?

**Comment by AaronBallman** [comment]
> > I don't think wanting the precommit CI repo to live within the project it tests is a reason to switch to github PRs (especially given that there is not clear consensus in the community to switch to GitHub PRs).
> 
> I thought the move to PRs was already decided and we’re just months away?

That would be news to me, but also (sadly) wouldn't surprise me with the way these decisions have happened in the past.

**Comment by joker-eph** [comment]
(Before this announcement I was planning on migrating Phabricator hosting to the foundation, but felt it wasn’t worth investing there and better focus on blockers for PRs, for me)

**Comment by AaronBallman** [comment]
> Here: I found it again https://discourse.llvm.org/t/code-review-process-update/63964

That was the thread I was thinking of as well; it is not at all clear (at least to me) that there is consensus on that thread.
